### PR TITLE
Move tasks in build scripts to task avoidance api

### PR DESCRIPTION
**Pull request by breskeby**
- Some trivial cleanup on build scripts
- Change task referencing in build scripts to use task avoidance api
where replacement is trivial.

Before:
https://gradle-enterprise.elastic.co/s/yqoztiubfz5za/performance/configuration#summary-task-totals

After:
https://gradle-enterprise.elastic.co/s/giqtxsr5z25d2/performance/configuration#summary-task-totals

### CODE REVIEW

File: modules/reindex/build.gradle [hunk 1]
```diff
@@ -146,7 +148,7 @@ if (Os.isFamily(Os.FAMILY_WINDOWS)) {
       env 'JAVA_HOME', jdks.legacy.javaHomePath
       args 'oldes.OldElasticsearch',
         baseDir,
-        unzip.temporaryDir,
+        unzip.get().temporaryDir,
```
Comment (mark-vieira) [inline modules/reindex/build.gradle:151]: Doesn't this effectively nullify any benefit of lazily creating the unzip task? Is there a reason we can't lazily create the fixture task?

File: plugins/repository-gcs/build.gradle [hunk 2]
```diff
@@ -293,29 +293,28 @@ testClusters {
...
-    setTestClassesDirs(yamlRestTestSourceSet.getOutput().getClassesDirs())
-    setClasspath(yamlRestTestSourceSet.getRuntimeClasspath())
-}
+  SourceSetContainer sourceSets = project.getExtensions().getByType(SourceSetContainer.class);
+  SourceSet yamlRestTestSourceSet = sourceSets.getByName(YamlRestTestPlugin.SOURCE_SET_NAME)
+  setTestClassesDirs(yamlRestTestSourceSet.getOutput().getClassesDirs())
+  setClasspath(yamlRestTestSourceSet.getRuntimeClasspath())
 
-check.dependsOn largeBlobYamlRestTest
+  // We have to wait for configure the cluster here as it might not have been created otherwise yet.
```
Comment (mark-vieira) [inline plugins/repository-gcs/build.gradle:306]: By this you just mean that we move this config _inside_ the task configuration block as that guarantees the task has been created, and subsequently, the corresponding testcluster will exist as well?

File: qa/wildfly/build.gradle [hunk 3]
```diff
@@ -63,8 +63,8 @@ elasticsearch_distributions {
   }
 }
 
-preProcessFixture {
-  dependsOn war, elasticsearch_distributions.docker
+tasks.named("preProcessFixture").configure {
+  dependsOn "war", elasticsearch_distributions.docker
```
Comment (mark-vieira) [inline qa/wildfly/build.gradle:67]: I see we wrap the task dependency in quotes here, but not later with `integTest`. In practice, does it matter? Obviously this config block is deferred until this task is needed, at which point we'll need all its dependencies as well, right? Essentially, there's no scenario in which this block will evaluate and we don't need to create the `war` task.

File: x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle [hunk 4]
```diff
@@ -11,7 +11,7 @@ dependencies {
 }
 
 task "leader-cluster"(type: RestIntegTestTask) {
-  mustRunAfter(precommit)
+  mustRunAfter("precommit")
```
Comment (mark-vieira) [inline x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle:14]: I feel like the rest test plugin or something should be adding this ordering rule.

**Review by mark-vieira — APPROVED** [review]
Some minor comments but otherwise LGTM.

File: buildSrc/build.gradle [hunk 5]
```diff
@@ -146,19 +148,29 @@ if (project != rootProject) {
   apply plugin: 'elasticsearch.publish'
 
   // groovydoc succeeds, but has some weird internal exception...
-  groovydoc.enabled = false
+  tasks.named("groovydoc").configure {
+    it.enabled = false
```
[the remark this answers is not in the document]
Comment (breskeby) [reply buildSrc/build.gradle:152]: I'm going ahead and remove implicit `it` where its not required

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy [hunk 6]
```diff
@@ -108,6 +108,21 @@ class PluginBuildPlugin implements Plugin<Project> {
             }
         }
 
+        // We've ported this from multiple build scripts where we see this pattern into
+        // an extension method as a first step of consolidation.
+        // We might want to port this into a general pattern later on.
+        project.ext.addQaCheckDependencies = {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy:114]: I agree. I just didn't want to change the behaviour with this PR. I'll put this on my todo list to clean up in another separate PR

File: buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy [hunk 7]
```diff
@@ -108,6 +108,21 @@ class PluginBuildPlugin implements Plugin<Project> {
...
+            project.afterEvaluate {
+                // let check depend on check tasks of qa sub-projects
+                def checkTaskProvider = project.tasks.named("check")
+                def qaSubproject = project.subprojects.find { it.path == project.path + ":qa" }
+                if(qaSubproject) {
+                    qaSubproject.subprojects.findAll {p -> p.path.startsWith(project.path + ":qa") }
```
[the remark this answers is not in the document]
Comment (breskeby) [reply buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy:120]: indeed.

File: modules/reindex/build.gradle (hunk 1 above)
Comment (breskeby) [reply modules/reindex/build.gradle:151]: yeah this just kicks the can down to the fixture task creation. Changing these fixtures to use task avoidance had some side effects in an earlier trial I want to look in an dedicated effort. Again this PR is basically of just porting the low hanging fruits

File: plugins/repository-gcs/build.gradle (hunk 2 above)
Comment (breskeby) [reply plugins/repository-gcs/build.gradle:306]: exactly

File: qa/wildfly/build.gradle (hunk 3 above)
Comment (breskeby) [reply qa/wildfly/build.gradle:67]: indeed. there's only odd scenarios where this makes a difference I think. (e.g.) materialising this task without the need of having it in the task graph (by accident during configuration time). IMO ideally this materialisation would only happen by gradle itself when calculation the taskgraph

File: x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle (hunk 4 above)
Comment (breskeby) [reply x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle:14]: you mean that tasks of type `RestIntegTestTask` should always run after `precommit` ?

Comment (mark-vieira) [reply x-pack/plugin/ccr/qa/downgrade-to-basic-license/build.gradle:14]: Yeah, I would think so.

**Pull request merged**